Add touch and controller support for :fly

ezekieltem opened the pull request:
Adds flight support for Touch/Mobile devices and Controllers

ezekieltem committed:
Update fly.lua

Added multi-platform support

TheNexusAvenger commented:
This diff is hard to understand because the entire file was converted from 4 spaces to indent to tabs. Can it be converted back to be consistent with the rest of the project?

ezekieltem committed:
Change indent formatting to 4 spaces

ezekieltem committed:
Remove 2 unneeded lines

Remove two ContextActionService:UnbindAction() calls that aren't needed.

TheNexusAvenger commented:
It may be a day or two before I test the functionality of this, but I wanted to comment now that I like how this was implemented. All that stands out to me is formatting that inconsistent with the rest of the project.

ezekieltem committed:
Clear up blank line differences

Hopefully clears all differences on blank lines where the only difference is spaces

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -70,6 +75,21 @@ return {
                         [Enum.KeyCode.RightShift] = false,
                     }
+                    -- Create a link for every InputBegan and InputEnded
+                    Flight.MultiPlatform = {
```
Can this become a constant instead of part of the `Flight` object?

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
                         end
                     end
+
+                    ContextActionService:BindAction("toggleFlight",function(actionName, UserInputState, InputObject)
```
Camel case isn't used in the project, and there should be a space after the comma. Requested line:
```suggestion
                    ContextActionService:BindAction("ToggleFlight", function(ActionName, UserInputState, InputObject)
```

(Also remember to change `toggleFlight` to `ToggleFlight` in other cases)

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
                         end
                     end))
+                    table.insert(Flight.Events, UserInputService.InputChanged:Connect(function(Key)
+                        local X,Y = nil,nil
```
Again, spaces after commas.
```suggestion
                        local X, Y = nil, nil
```

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
+                        else
+                            (Flight :: any):Start()
+                        end
+                    end,true,Enum.KeyCode.E,Enum.KeyCode.ButtonX)
+                    ContextActionService:SetTitle("toggleFlight","Toggle Flight")
+                    ContextActionService:SetPosition("toggleFlight",UDim2.new(0,50,0,100))
```
These 3 lines should have spaces after the commas. Requested lines:
```suggestion
                    end, true, Enum.KeyCode.E, Enum.KeyCode.ButtonX)
                    ContextActionService:SetTitle("ToggleFlight", "Toggle Flight")
                    ContextActionService:SetPosition("ToggleFlight", UDim2.new(0, 50, 0, 100))
```

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
                         end
+
+                        --Unbind the ContextActionService bind
+                        ContextActionService:UnbindAction("toggleFlight")
```
Like the other, changing the camel case to pascal case.
```suggestion
                        ContextActionService:UnbindAction("ToggleFlight")
```

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -10,14 +10,15 @@ local Players = game:GetService("Players")
     Keyword = "fly",
     Category = "UsefulFunCommands",
-    Description = "Gives a set of players the ability to fly. Use E to toggle on/off.",
+    Description = "Gives a set of players the ability to fly. Use E on keyboard or "..UserInputService:GetStringForKeyCode(Enum.KeyCode.ButtonX):sub(7).." on controller to toggle on/off.",
```
The rest of the project uses `string.sub`. I'd like it used for consistency.
```suggestion
    Description = "Gives a set of players the ability to fly. Use E on keyboard or "..string.sub(UserInputService:GetStringForKeyCode(Enum.KeyCode.ButtonX), 7).." on controller to toggle on/off.",
```

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
+                        end
+
+                        if X and Y then
+                            if X <= -.5 then
+                                Flight.KeysDown[Enum.KeyCode.A] = true
+                                Flight.KeysDown[Enum.KeyCode.D] = false
+                            elseif X >= .5 then
+                                Flight.KeysDown[Enum.KeyCode.A] = false
+                                Flight.KeysDown[Enum.KeyCode.D] = true
+                            else
+                                Flight.KeysDown[Enum.KeyCode.A] = false
+                                Flight.KeysDown[Enum.KeyCode.D] = false
+                            end
+
+                            if Y <= -.5 then
+                                Flight.KeysDown[Enum.KeyCode.S] = true
+                                Flight.KeysDown[Enum.KeyCode.W] = false
+                            elseif Y >= .5 then
```
All other decimals in the project include the leading zero. Please include it for consistency.
```suggestion
                            if X <= -0.5 then
                                Flight.KeysDown[Enum.KeyCode.A] = true
                                Flight.KeysDown[Enum.KeyCode.D] = false
                            elseif X >= 0.5 then
                                Flight.KeysDown[Enum.KeyCode.A] = false
                                Flight.KeysDown[Enum.KeyCode.D] = true
                            else
                                Flight.KeysDown[Enum.KeyCode.A] = false
                                Flight.KeysDown[Enum.KeyCode.D] = false
                            end

                            if Y <= -0.5 then
                                Flight.KeysDown[Enum.KeyCode.S] = true
                                Flight.KeysDown[Enum.KeyCode.W] = false
                            elseif Y >= 0.5 then
```

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
+                    table.insert(Flight.Events,Flight.Humanoid:GetPropertyChangedSignal("MoveDirection"):Connect(function()
+                        -- Stop the flight movement when the Humanoid has no Movement Direction
+                        local MoveDirection = Flight.Humanoid.MoveDirection
+                        if MoveDirection.X < .5 and MoveDirection.X > -.5 and MoveDirection.Z < .5 and MoveDirection.Z > -.5 then
```
Again, leading zeros for decimals.
```suggestion
                        if MoveDirection.X < 0.5 and MoveDirection.X > -0.5 and MoveDirection.Z < 0.5 and MoveDirection.Z > -0.5 then
```

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -247,42 +267,101 @@ return {
                     end))
-    
+
+                    table.insert(Flight.Events,Flight.Humanoid:GetPropertyChangedSignal("MoveDirection"):Connect(function()
```
Again, spaces after commas.
```suggestion
                    table.insert(Flight.Events, Flight.Humanoid:GetPropertyChangedSignal("MoveDirection"):Connect(function()
```

TheNexusAvenger reviewed (commented):
The functionality is good, but there are some formatting changes I would like to see before merging.

This is also my first time using the suggestions feature. I hope I didn't mess anything up.

ezekieltem committed:
change :sub to string.sub

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
change toggleFlight to ToggleFlight

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Update 2nd use of toggleFlight to ToggleFlight; add space after comma

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
more toggleFlight to ToggleFlight; and more spaces after commas

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Update src/IncludedCommands/UsefulFun/fly.lua

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Update src/IncludedCommands/UsefulFun/fly.lua

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Update src/IncludedCommands/UsefulFun/fly.lua

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Update src/IncludedCommands/UsefulFun/fly.lua

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Change Flight.MultiPlatform to be constant

ezekieltem commented:
Think I got all the suggestions, let me know if there were any I missed.

TheNexusAvenger commented on src/IncludedCommands/UsefulFun/fly.lua:
```diff
@@ -10,14 +10,31 @@ local Players = game:GetService("Players")
 local Types = require(script.Parent.Parent.Parent:WaitForChild("Types"))
 
+-- Create a link for every InputBegan and InputEnded
+local MultiPlatform = {
+
+    -- Keyboard
```
One last formatting change and it should be done: for constants, snake case with caps locks is used (ex: `MULTI_PLATFORM`) is used. I'd also like to request the name be a bit more descriptive than that, like `MULTIPLATFORM_KEYCODE_MAPPING`.

```suggestion
local MULTIPLATFORM_KEYCODE_MAPPING= {
    -- Keyboard
```

TheNexusAvenger reviewed (commented):
1 last formatting request before I accept it.

ezekieltem committed:
Update src/IncludedCommands/UsefulFun/fly.lua

Co-authored-by: TheNexusAvenger <13441476+TheNexusAvenger@users.noreply.github.com>

ezekieltem committed:
Correct old MultiPlatform uses to MULTIPLATFORM_KEYCODE_MAPPING

TheNexusAvenger reviewed (approved):
The changes look good to me. Thanks for contributing this long-requested feature.
